### PR TITLE
[expo][keep-awake] Fix `Unable to deactivate keep awake` warnings

### LINKED ISSUE #12584: Error: Unable to deactivate keep awake. However, it probably is deactivated already.
### Summary

Once I launch the app in the Android emulator and If I press the back button(triangle shaped) and then press the Square shaped button on the Android emulator and try to relaunch the app I see the following error message.

Possible Unhandled Promise Rejection (id: 0): Error: Unable to deactivate keep awake. However, it probably is deactivated already. 

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

_No response_

### Environment

  Expo CLI 4.4.1 environment info:
    System:
      OS: macOS 11.2.3
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 14.15.4 - /usr/local/bin/node
      npm: 7.7.6 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.4, DriverKit 20.2, macOS 11.1, tvOS 14.3, watchOS 7.2
      Android SDK:
        API Levels: 29, 30
        Build Tools: 28.0.3, 29.0.2, 30.0.3
        System Images: android-28 | Intel x86 Atom_64, android-28 | Google APIs Intel x86 Atom, android-28 | Google APIs Intel x86 Atom_64, android-28 | Google Play Intel x86 Atom, android-28 | Google Play Intel x86 Atom_64, android-29 | Google APIs Intel x86 Atom, android-29 | Google Play Intel x86 Atom, android-29 | Google Play Intel x86 Atom_64, android-30 | Google APIs Intel x86 Atom, android-30 | Google Play ARM 64 v8a, android-30 | Google Play Intel x86 Atom, android-30 | Google Play Intel x86 Atom_64
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.7042882
      Xcode: 12.4/12D4e - /usr/bin/xcodebuild
    npmPackages:
      expo: ~41.0.0 => 41.0.0 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: ~0.63.4 => 0.63.4 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.4.1
    Expo Workflow: bare

### Reproducible demo or steps to reproduce from a blank project

1) Create a new app by using the following command.
npx create-react-native-app notification
2) Following command is used to launch the app.
react-native run-android
3) After app is launched, press back button(Triangle shaped) on android emulator.
4) Now press the Sqaure shaped button on the emulator and launch the app again.

### PR DESCRIPTION
**Pull request by Kudo**
# Why

a slightly different proposal of #13432 and to fix #12584

my hypothesis is that in the case, the Activity is dead (`finish()` called) but the process is still alive and the react-native bridge is still alive. before the Activity dead, `deactivateKeepAwake` is called. however, the unhandled promise rejection will be kept in the bridge queue. it will show in the yellow box when next time the new Activity created.

sometimes, android will not finish Activity when pressing back button, [especially on android 12](https://developer.android.com/about/versions/12/behavior-changes-all#back-press). to reliably test the issue, i would turn on the `Don't keep activities` flag in android developer menu:

according to the study from #13432, the `FLAG_KEEP_SCREEN_ON` scope is window based. if the Activity is dead, we don't need to remove the flag.

# How

based on the observations above, my assumption is that it's a common case and we don't need to show the warning. this pr is to remove the warning.

- update `activateKeepAwake` and `deactivateKeepAwake` to return `Promise<void>`, so that we can handle to promise rejection.
- add `{ suppressDeactivateWarnings: true }` option to suppress the promise rejection.
- use the `{ suppressDeactivateWarnings: true }` in `withExpoRoot`

# Test Plan

#### Integration Test

1. turn on `Don't keep activities` flag in android developer menu.
2. on a sdk45 project
3. launch app -> press back (triangle) button -> press square button to resume the app. see whether the warning is displayed.

#### Unit Test

```
  useKeepAwake
    ✓ test default calls without any parameters (6 ms)
    ✓ test explicit calls with parameters (1 ms)
```

i did try to test the promise rejection. i got no luck to test the promise rejection in hook as well as the `useEffect` cleanup function.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]

File: packages/expo-keep-awake/src/__tests__/index-test.native.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  });
+
+  it('test explicit calls with parameters', async () => {
+    const { unmount } = renderHook(() => useKeepAwake('tag', { suppressDeactivateWarnings: true }));
+    unmount();
+    expect(mockActivate.mock.calls.length).toBe(1);
+    expect(mockDeactivate.mock.calls.length).toBe(1);
+    expect(mockDeactivate.mock.calls[0][0]).toEqual('tag');
+  });
+});
```
Comment (byCedric) [inline packages/expo-keep-awake/src/__tests__/index-test.native.ts:35]: Maybe also nice to add a test case to validate the suppressing mechanism? e.g.

```suggestion
  it('supresses uncaught exceptions', async () => {
    mockDeactivate.mockRejectedValue(new Error('unable to deactivate');
    const errorSpy = jest.spyOn(console, 'error');
    const { unmount } = renderHook(() => useKeepAwake('tag', { suppressDeactivateWarnings: true }));
    unmount();
    expect(errorSpy).not.toBeCalled();
  });
});
```

(no idea if this test works, I freestyled it)

**Review by byCedric — APPROVED** [review]
gentleman-club

**Review by tsapeta — APPROVED** [review]

**Review by bbarthec — APPROVED** [review]

**Comment by Kudo** [comment]
noticed the ios test suite ci job broken from my accidental copied change. added the last change to fix it 🤞

**Comment by Kudo** [comment]
the ios ci jobs failure may come from pod cache with [this commit](https://github.com/expo/expo/commit/ccbd02e04c350590540b09e93d37b6ebb4b1fe4d) on main branch. i'll merge this pr anyway and open another pr to fix the ios failure.

**Comment by siarheipashkevich** [comment]
image

On the activation we also have the problem (

**Comment by Kudo** [comment]
@siarheipashkevich would be happy to investigate if you could have a minimal reproducible example.

**Comment by siarheipashkevich** [comment]
@Kudo it's very simple we should switch to another tab before the page is loaded. And when I back to tab with my app I saw this error, I think it's because the page is not loaded.. if you need I can record video

**Comment by Kudo** [comment]
@siarheipashkevich having these steps and it works for me. so it would be good if you could create a reproducible project for us to investigate.

```sh
$ yarn create expo-app -t tabs@sdk-48 sdk48
$ cd sdk48
$ npx expo start -w
```

**Comment by siarheipashkevich** [comment]
@Kudo https://github.com/siarheipashkevich/expo-keep-awake

**Pull request merged**